### ISSUE TITLE
Silently incorrect matmul precision when provided as strings on GPUs

### ISSUE THREAD

**Issue by shoyer** [jax-ml/jax#39869]
The legacy precision strings (e.g., `jnp.matmul(..., precision='bfloat16')`) are aliases for the 3-level `lax.Precision` enum: `'bfloat16_3x'` and `'tensorfloat32'` both map to `Precision.HIGH`, and `'bfloat16'` maps to `Precision.DEFAULT`: https://github.com/jax-ml/jax/blob/4f313089c00e13600eb918bf47919c252b4a1fad/jax/_src/lax/lax.py#L2175-L2182

These names are unfortunately only correct on TPUs, because XLA understands `Precision.DEFAULT` and `Precision.HIGH` as `tensorfloat32` on GPUs.  Note that the named algorithms *are* available on GPU via `lax.DotAlgorithmPreset` (e.g. `BF16_BF16_F32_X3`), but the strings don't select them.

The current situation is a poor user experience for GPU users in two different ways:
1. With `precision=bfloat16_3x`, users encounter unexpected loss of precision, because `tensorfloat32` has less precision than `bfloat16_3x` (e.g., https://github.com/jax-ml/jax/issues/19974, https://github.com/neuralgcm/dinosaur/pull/138)
2. With `precision='bfloat16'`, users encounter unexpected slow-down, because `tensorfloat32` is slower than `bfloat16`.

`dtype='float32'` and `dtype='tensorfloat32'` are also incorrect on TPUs, but these aliases are _somewhat_ less problematic, because TPUs lack native support for these dtypes, so there is no loss of performance and `HIGHEST`/`HIGH` map to nearly equivalent or higher precision (technically `HIGHEST` has less precision than full float32, but 6x passes of bfloat16 is equivalent except for denormals).

This behavior is long-standing, but it's problematic enough that I think it would be appropriate to raise an error in these cases (i.e., `precision='bfloat16_3x'` and `precision='bfloat16'` when run on GPUs). At the very least JAX should issue loud warnings.

Alternatively, these strings could be updated to actually implement the correct algorithm on GPUs, but I think it's too late for that now. Mismatched JAX versions would cause silent loss of precision or slow down. It is better to encourage switching to an unambiguous `DotAlgorithm`.

CC @jakevdp 

## Repro (A100, jax 0.11.0, CUDA 12)
<details>
```python
import functools
import numpy as np
import jax
import jax.numpy as jnp

rng = np.random.default_rng(0)
a = rng.standard_normal((2048, 2048)).astype(np.float32)
b = rng.standard_normal((2048, 2048)).astype(np.float32)
exact = np.matmul(a.astype(np.float64), b.astype(np.float64))
scale = np.abs(exact).max()

for precision in ['bfloat16', 'tensorfloat32', 'bfloat16_3x', 'float32',
                  jax.lax.DotAlgorithmPreset.TF32_TF32_F32,
                  jax.lax.DotAlgorithmPreset.TF32_TF32_F32_X3,
                  jax.lax.DotAlgorithmPreset.BF16_BF16_F32,
                  jax.lax.DotAlgorithmPreset.BF16_BF16_F32_X3]:
  f = jax.jit(functools.partial(jnp.matmul, precision=precision))
  err = np.abs(np.asarray(f(a, b)) - exact).max() / scale
  print(f'{str(precision):<40} max rel err {err:.3e}')
```

```
bfloat16                                 max rel err 2.799e-04
tensorfloat32                            max rel err 2.799e-04
bfloat16_3x                              max rel err 2.799e-04
float32                                  max rel err 2.092e-06
TF32_TF32_F32                            max rel err 2.799e-04
TF32_TF32_F32_X3                         max rel err 7.787e-06
BF16_BF16_F32                            max rel err 2.345e-03
BF16_BF16_F32_X3                         max rel err 1.421e-05
```

The first three strings all produce errors identical to `TF32_TF32_F32`, i.e. single-pass TF32 (~10-bit mantissa), rather than the ~`2^-21`-accurate 3-pass bfloat16 algorithm that `'bfloat16_3x'` names (and runs on TPU).
</details>

**Comment by jakevdp** [comment]
Thanks for raising this. What would you think about deprecating string inputs altogether? It seems like using the existing enums would lead to less ambiguity.

**Comment by mattjj** [comment]
(Great to hear from you @shoyer!)

Yeah I tend to think deprecating them would be ideal, if we can pull it off.

Slightly tangential, but here are [some new docs](https://docs.jax.dev/en/latest/new_docs/201/precision.html) about precisions. Criticisms welcome.

**Comment by shoyer** [comment]
I can see a case for platform dependent presets like fastest/highest (i.e., "use me for neural nets" vs "use for me scientific computing"), but if they're going to be named like that they should at least be honest, i.e., "fastest" should be bfloat16 rather than tensorfloat32 on modern GPUs. But if we're going to make aliases like that now, they would need to be new names that are accurately advertised/documented from the start.

So yes, I would definitely support deprecating the old precision strings across the board. I do think it's important to highlight, at least via customized warning messages, that some cases (namely, bfloat16 and bfloat16_3x on GPU) can they have silent degraded accuracy or performance. We don't want users to just silence warnings about new APIs in that case.

@mattjj My main suggested addition to these lovely new docs would be to note that dot precision can be shape dependent, because `jnp.dot` can be a vector-vector or matrix-vector product that always use tensor cores. It would be good to explain how XLA handles these cases, and note that specifying precision can still be a good idea if anticipating `vmap`.
